File: Dsa-day149.py
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        order = {"electronics": 0, "grocery": 1, "pharmacy": 2, "restaurant": 3}
        candidates = []
        for i in range(len(code)):
            if (code[i] and 
                isActive[i] and 
                businessLine[i] in order and 
                all(c.isalnum() or c == '_' for c in code[i])):
                candidates.append((order[businessLine[i]], code[i]))
        candidates.sort()
        return [code_val for _, code_val in candidates]

File: Dsa-day149.cpp
class Solution {
public:
    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        unordered_map<string, int> order = {
            {"electronics", 0},
            {"grocery", 1},
            {"pharmacy", 2},
            {"restaurant", 3}
        };
        vector<pair<int, string>> candidates;
        int n = code.size();
        for (int i = 0; i < n; ++i) {
            const string& c = code[i];
            const string& bl = businessLine[i];
            if (!c.empty() && isActive[i] && order.count(bl)) {
                bool valid_code = true;
                for (char ch : c) {
                    if (!isalnum(ch) && ch != '_') {
                        valid_code = false;
                        break;
                    }
                }
                if (valid_code) {
                    candidates.emplace_back(order[bl], c);
                }
            }
        }
        sort(candidates.begin(), candidates.end());
        vector<string> result;
        for (const auto& p : candidates) {
            result.push_back(p.second);
        }
        return result;
    }
};